src/rgw/rgw_rest_s3.cc

```diff
@@ -6494,6 +6494,7 @@ rgw::auth::s3::LocalEngine::authenticate(
 
   /* Ignore signature for HTTP OPTIONS */
   if (s->op_type == RGW_OP_OPTIONS_CORS) {
+    s->user->set_attrs(user->get_attrs());
     auto apl = apl_factory->create_apl_local(
         cct, s, user->get_info(), std::move(account), std::move(policies),
```

Comment:
these auth engines get `const req_state*` intentionally to prevent them from making changes like this to the request. the fact that `s->user` is a pointer makes it possible to work around that restriction, but i'd prefer a design that doesn't

we're already passing `user->get_info()` in so that it can be copied into the request through `IdentityApplier::load_acct_info(const DoutPrefixProvider* dpp, RGWUserInfo& user_info)`. we call that [in a single place](https://github.com/ceph/ceph/blob/d908c9e/src/rgw/rgw_auth.cc#L524-L526) just after `authenticate()` succeeds:
```c++
      applier->load_acct_info(dpp, s->user->get_info());
```

i think the right thing to do is pass our initialized `user` (instead of just the `user->get_info()`) into `create_apl_local()`, and expand `load_acct_info()` so that it fully initializes `s->user`. for example,
```c++
virtual auto load_acct_info(const DoutPrefixProvider* dpp)
  -> std::unique_ptr<rgw::sal::User> = 0;
```
and
```c++
      s->user = applier->load_acct_info(dpp);
```

Reply:
@cbodley hi, pushed an update - unfortunately the following did not work out:
> ```c++
> virtual auto load_acct_info(const DoutPrefixProvider* dpp)
>   -> std::unique_ptr<rgw::sal::User> = 0;
> ```  
> and
> ```c++
>       s->user = applier->load_acct_info(dpp);
> ```
because there are complication steaming from use cases in which the passed `RGWUserInfo& user_info` parameter is being acted upon so was not able to elide it.

https://github.com/ceph/ceph/blob/505f1a67a9238cdf4845a4f4b5bdbeeaf0f908c4/src/rgw/rgw_auth.cc#L634C1-L645C31